### ISSUE TITLE
warp_policy will make GPU hang when training mixed multi modal data

### ISSUE THREAD

**Issue by techkang** [volcengine/verl#3491]
### System Info

----------Python Info----------
Version      : 3.10.12
Compiler     : GCC 11.4.0
Build        : ('main', 'Jul 29 2024 16:56:48')
Arch         : ('64bit', 'ELF')
------------Pip Info-----------
Version      : 25.1.1
Directory    : /usr/local/lib/python3.10/dist-packages/pip
vllm         : 0.10.0
sglang       : not found.
ray          : 2.47.1
torch        : 2.7.1
----------verl Info-----------
Version      : 0.5.0.dev
Directory    : /file_system/kangsheng/verl/verl
Commit Hash  : 44cdc4a5275c77c2f64487885c1680997d6cbab1
----------Platform Info----------
Platform     : Linux-5.4.250-4-velinux1u1-amd64-x86_64-with-glibc2.35
system       : Linux
node         : di-20250904113923-94xhn
release      : 5.4.250-4-velinux1u1-amd64
version      : #velinux1u1 SMP Debian 5.4.250-4-velinux1u1 Thu May 23 09:30:06 
----------Environment----------
CUDA Runtime : 12.6
CUDA Compiler : Cuda compilation tools, release 12.6, V12.6.20
----------System Info----------
CPU Memory      : 1785.00 GB
GPU Count       : 8
GPU 1   Type    : NVIDIA H20
GPU 1   Memory  : 95.58 GB
GPU 2   Type    : NVIDIA H20
GPU 2   Memory  : 95.58 GB
GPU 3   Type    : NVIDIA H20
GPU 3   Memory  : 95.58 GB
GPU 4   Type    : NVIDIA H20
GPU 4   Memory  : 95.58 GB
GPU 5   Type    : NVIDIA H20
GPU 5   Memory  : 95.58 GB
GPU 6   Type    : NVIDIA H20
GPU 6   Memory  : 95.58 GB
GPU 7   Type    : NVIDIA H20
GPU 7   Memory  : 95.58 GB
GPU 8   Type    : NVIDIA H20
GPU 8   Memory  : 95.58 GB

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

minimal scripts to reproduce (uncomment `warp_policy = get_fsdp_wrap_policy(self.model, {'min_num_params': 0})` and run script, the code will hang)
```
import torch
import torch.distributed as dist
from torch.nn.parallel import DistributedDataParallel as DDP
from torch.distributed.fsdp import FullyShardedDataParallel as FSDP, ShardingStrategy
from torch.distributed.fsdp.wrap import (
    size_based_auto_wrap_policy,
    enable_wrap,
    wrap,
)
import functools
import ray
from PIL import Image
import numpy as np
from transformers import Qwen2_5_VLForConditionalGeneration, AutoProcessor
import os
from verl.utils.fsdp_utils import get_fsdp_wrap_policy

from qwen_vl_utils import process_vision_info

@ray.remote(num_gpus=1)
class MultiModalWorker:
    def __init__(self, rank, world_size, master_addr="localhost", master_port="12355"):
        self.rank = rank
        print(f'self rank: {rank}')
        self.world_size = world_size
        
        # Initialize distributed environment
        self._init_distributed(rank, world_size, master_addr, master_port)
        
        # Set the correct CUDA device for this rank (map to available GPUs)
        available_gpus = torch.cuda.device_count()
        device_id = rank % available_gpus  # Cycle through available GPUs
        torch.cuda.set_device(device_id)
        self.device_id = device_id
        
        # Load model and processor
        self.model = Qwen2_5_VLForConditionalGeneration.from_pretrained(
            "Qwen/Qwen2.5-VL-7B-Instruct",
            torch_dtype=torch.bfloat16,
            attn_implementation="flash_attention_2",
            device_map=None
        )
        self.model = self.model.to(f'cuda:{device_id}')
        
        # warp_policy = get_fsdp_wrap_policy(self.model, {'min_num_params': 0})
        warp_policy = None
        self.model = FSDP(
            self.model,
            mixed_precision=None,
            auto_wrap_policy=warp_policy,
            device_id=device_id,
        )
        # self.model = DDP(
        #     self.model,
        #     find_unused_parameters=True
        # )
        
        self.processor = AutoProcessor.from_pretrained("Qwen/Qwen2.5-VL-7B-Instruct")
        
        # Optimizer
        self.optimizer = torch.optim.SGD(self.model.parameters(), lr=1e-4)
        
        print(f"Worker {rank} initialized on cuda:{device_id} with FSDP")
    
    def _init_distributed(self, rank, world_size, master_addr, master_port):
        """Initialize distributed process group"""
        os.environ['MASTER_ADDR'] = master_addr
        os.environ['MASTER_PORT'] = master_port
        os.environ['RANK'] = str(rank)
        os.environ['WORLD_SIZE'] = str(world_size)
        os.environ['LOCAL_RANK'] = str(rank)
        
        # Initialize the process group
        dist.init_process_group(
            backend='nccl',
            rank=rank,
            world_size=world_size
        )
        
        print(f"Initialized distributed training: rank {rank}/{world_size}")
    
    def generate_data(self, idx):
        """Generate synthetic training data based on rank"""
        if self.rank % 2 == 0:  # text_only
            messages = [{
                "role": "user", 
                "content": [{"type": "text", "text": f"Random text sample {idx}"}]
            }]
        else:  # image_text
            image = Image.fromarray(np.random.randint(0, 255, (1024, 1024, 3), dtype=np.uint8))
            messages = [{
                "role": "user",
                "content": [
                    {"type": "image", "image": image},
                    {"type": "text", "text": f"Describe this image {idx}"}
                ]
            }]

        # Process inputs
        text = self.processor.apply_chat_template(messages, tokenize=False, add_generation_prompt=True)
        image_inputs, video_inputs = process_vision_info(messages)
        

        inputs = self.processor(
            text=[text],
            images=image_inputs, 
            videos=video_inputs,
            padding=True,
            return_tensors="pt"
        )
        
        
        # Move to device
        for key in inputs:
            if inputs[key] is not None:
                inputs[key] = inputs[key].to(f'cuda:{self.device_id}')
        
        # Create labels
        labels = inputs.input_ids.clone()
        labels[:, :-1] = inputs.input_ids[:, 1:]
        labels[:, -1] = -100
        inputs['labels'] = labels
        
        return inputs
    
    
    @torch.no_grad()
    def train_step(self, batch_indices):
        """Training step with gradient synchronization via DDP"""
        self.model.train()
        total_loss = 0.0
        num_samples = 0
        
        for idx in batch_indices:
            inputs = self.generate_data(idx)
            
            # Forward pass
            print(f"[Rank {self.rank}], {list(inputs.keys())=}")
            
            with torch.autocast(device_type='cuda', dtype=torch.bfloat16):
                outputs = self.model(**inputs)
            loss = outputs.loss
            print(f"  Training loss: {loss.item():.4f}")
            
            # # Synchronize after forward pass
            if dist.is_initialized():
                print(f"[Rank {self.rank}] Entering barrier after forward pass for idx {idx}")
                dist.barrier()
                print(f"[Rank {self.rank}] Exited barrier after forward pass for idx {idx}")
            
        
        avg_loss = total_loss / len(batch_indices) if batch_indices else 0.0
        
        # Synchronize loss across all workers for consistent reporting
        if dist.is_initialized():
            loss_tensor = torch.tensor(avg_loss, device=f'cuda:{self.device_id}')
            dist.all_reduce(loss_tensor, op=dist.ReduceOp.AVG)
            synchronized_loss = loss_tensor.item()
        else:
            synchronized_loss = avg_loss
            
        return synchronized_loss, self.rank
    
    def cleanup(self):
        """Cleanup distributed process group"""
        if dist.is_initialized():
            dist.destroy_process_group()
            print(f"Worker {self.rank} destroyed process group")
    
    def get_model_state(self):
        """Get model state dict for synchronization"""
        return self.model.state_dict()
    
    def set_model_state(self, state_dict):
        """Set model state dict for synchronization"""
        self.model.load_state_dict(state_dict)
    
    def get_dataset_size(self):
        """Get dataset size - return fixed size for synthetic data"""
        return 1000

def average_model_states(state_dicts):
    """Average model parameters across workers"""
    if not state_dicts:
        return None
    
    # Get the first state dict as template
    avg_state = {}
    for key in state_dicts[0].keys():
        # Average the parameters
        avg_state[key] = torch.stack([state[key] for state in state_dicts]).mean(dim=0)
    
    return avg_state

def main():
    # Initialize Ray
    ray.init()
    
    # Check available GPUs
    num_gpus = torch.cuda.device_count()
    num_workers = num_gpus  # Reduce workers for easier distributed setup
    
    print(f"Available GPUs: {num_gpus}, Using workers: {num_workers}")
    
    # Create workers with distributed parameters
    workers = []
    master_addr = "localhost"
    master_port = "12355"
    
    for i in range(num_workers):
        worker = MultiModalWorker.remote(
            rank=i, 
            world_size=num_workers, 
            master_addr=master_addr, 
            master_port=master_port
        )
        workers.append(worker)
    
    # Wait for all workers to initialize distributed environment
    ray.get([worker.get_dataset_size.remote() for worker in workers])
    print("All workers initialized with distributed training")
    
    # Training parameters
    total_samples = 1000
    epochs = 3
    batch_size = 8  # Fixed batch size for easier distributed training
    
    print(f"Starting Ray-based distributed training with {num_workers} workers")
    
    for epoch in range(epochs):
        print(f"\nEpoch {epoch + 1}/{epochs}")
        
        # Calculate data distribution
        samples_per_worker = total_samples // num_workers
        batches_per_worker = samples_per_worker // batch_size
        
        print(f"Samples per worker: {samples_per_worker}, Batches per worker: {batches_per_worker}")
        
        # Training phase with gradient synchronization
        print("=== Training Phase ===")
        epoch_losses = []
        
        for batch_idx in range(batches_per_worker):
            # Create batch indices for each worker
            train_futures = []
            
            for worker_id, worker in enumerate(workers):
                start_idx = worker_id * samples_per_worker + batch_idx * batch_size
                end_idx = start_idx + batch_size
                batch_indices = list(range(start_idx, min(end_idx, (worker_id + 1) * samples_per_worker)))
                
                if batch_indices:
                    future = worker.train_step.remote(batch_indices)
                    train_futures.append(future)
            
    
    # Cleanup
    print("\nCleaning up distributed workers...")
    ray.get([worker.cleanup.remote() for worker in workers])
    
    print("Distributed training completed!")
    ray.shutdown()

if __name__ == "__main__":
    main()
```

### Expected behavior

run normally

**Comment by techkang** [comment]
A possible solution could be: https://github.com/hiyouga/EasyR1/blob/main/verl/utils/fsdp_utils.py#L52-L66
Do you have any ideas? @vermouth1992 @PeterSH6

**Comment by wlhgtc** [comment]
With the help of Gemini, it appears the issue may be caused by this specific code of [Qwen2.5 HF model](https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py#L1256).

> When a rank get pure text batch, it will skips the computation of the `VisionBlock`, FSDP on that rank waits forever for the gradients from those modules, causing it to hang. 

Based on his advice, I added some code after this section to address the problem, but the error persists.
```python
  # add a dummy forward to init image projection layer
  if pixel_values is None and pixel_values_videos is None and self.training:
      dummy_pixel_values = torch.zeros(
          (4, 2 * 3 * 14 * 14), device=inputs_embeds.device, dtype=inputs_embeds.dtype
      )
      dummy_image_grid_thw = torch.tensor([[1, 2, 2]], device=inputs_embeds.device)
      _ = self.visual(dummy_pixel_values, dummy_image_grid_thw)
```

**Comment by wlhgtc** [comment]
@hiyouga  Easy RL works well on mixed multi modal data, I got the same `wrap policy` in console.
I wonder how you make it works?
```python
# Easy RL
FSDP wrap policy: functools.partial(<function transformer_auto_wrap_policy at 0x7fafab5120e0>, transformer_layer_cls={<class 'transformers.models.qwen2_5_vl.modeling_qwen2_5_vl.Qwen2_5_VLDecoderLayer'>, <class 'transformers.models.qwen2_5_vl.modeling_qwen2_5_vl.Qwen2_5_VLVisionBlock'>}).

# verl
rap_policy: functools.partial(<function _or_policy at 0x7f10dd823910>, policies=[functools.partial(<function transformer_auto_wrap_policy at 0x7f10dd8237f0>, transformer_layer_cls={<class 'transformers.models.qwen2_5_vl.modeling_qwen2_5_vl.Qwen2_5_VLDecoderLayer'>, <class 'transformers.models.qwen2_5_vl.modeling_qwen2_5_vl.Qwen2_5_VLVisionBlock'>})])
```

**Comment by hiyouga** [comment]
@wlhgtc will be fixed in #3496
